File: 1-array/burst-balloons.py
"""
LeetCode 452. Minimum Number of Arrows to Burst Balloons
Technique: Sorting
Time Complexity: O(n log n)
Space Complexity: O(1)
"""


class Solution(object):
    def findMinArrowShots(self, points):
        points.sort()
        arrow = 1
        overlap = points[0][1]
        for i in range(1, len(points)):
            if points[i][0] > overlap:
                arrow += 1
                overlap = points[i][1]
            else:
                overlap = min(points[i][1], overlap)
        return arrow

File: 1-array/burst-balloons.js
/**
 * @param {number[][]} points
 * @return {number}
 */
var findMinArrowShots = function(points) {
  points.sort((pointA, pointB) => pointA[0] - pointB[0])
  let overlappingEnd = points[0][1]
  let arrow = 1
  for (let i=1; i < points.length; i++) {
    if (points[i][0] <= overlappingEnd) {
        overlappingEnd = Math.min(overlappingEnd, points[i][1])
    } else {
        arrow++;
        overlappingEnd = points[i][1]
    }
  }
  return arrow;
};